src/uu/pgrep/src/process.rs

```diff
@@ -0,0 +1,468 @@
+        let path = value.to_str().ok_or(())?;
+
+        let f = |prefix: &str| {
+            value
+                .iter()
+                .last()?
+                .to_str()?
+                .strip_prefix(prefix)?
+                .parse::<u64>()
+                .ok()
+        };
+
+        if path.contains("ttyS") {
+            // Case 2
+            f("ttyS").ok_or(()).map(TerminalType::TtyS)
+        } else if path.contains("tty") {
+            // Case 3
+            f("tty").ok_or(()).map(TerminalType::Tty)
+        } else {
+            Err(())
+        }
+    }
+}
+
+/// State or process
+#[derive(Debug, PartialEq, Eq)]
+pub enum RunState {
+    ///`R`, running
+    Running,
+    ///`S`, sleeping
+    Sleeping,
+    ///`D`, sleeping in an uninterruptible wait
+    UninterruptibleWait,
+    ///`Z`, zombie
+    Zombie,
+    ///`T`, traced or stopped
+    Stopped,
+}
+
+impl Display for RunState {
+    fn fmt(&self, f: &mut Formatter) -> fmt::Result {
+        match self {
+            RunState::Running => write!(f, "R"),
+            RunState::Sleeping => write!(f, "S"),
+            RunState::UninterruptibleWait => write!(f, "D"),
+            RunState::Zombie => write!(f, "Z"),
+            RunState::Stopped => write!(f, "T"),
+        }
+    }
+}
+
+impl TryFrom<char> for RunState {
+    type Error = io::Error;
+
+    fn try_from(value: char) -> Result<Self, Self::Error> {
+        match value {
+            'R' => Ok(RunState::Running),
+            'S' => Ok(RunState::Sleeping),
+            'D' => Ok(RunState::UninterruptibleWait),
+            'Z' => Ok(RunState::Zombie),
+            'T' => Ok(RunState::Stopped),
+            _ => Err(io::ErrorKind::InvalidInput.into()),
+        }
+    }
+}
+
+impl TryFrom<&str> for RunState {
+    type Error = io::Error;
+
+    fn try_from(value: &str) -> Result<Self, Self::Error> {
+        if value.len() != 1 {
+            return Err(io::ErrorKind::InvalidInput.into());
+        }
+
+        RunState::try_from(
+            value
+                .chars()
+                .nth(0)
+                .ok_or::<io::Error>(io::ErrorKind::InvalidInput.into())?,
+        )
+    }
+}
+
+impl TryFrom<String> for RunState {
+    type Error = io::Error;
+
+    fn try_from(value: String) -> Result<Self, Self::Error> {
+        RunState::try_from(value.as_str())
+    }
+}
+
+impl TryFrom<&String> for RunState {
+    type Error = io::Error;
+
+    fn try_from(value: &String) -> Result<Self, Self::Error> {
+        RunState::try_from(value.as_str())
+    }
+}
+
+/// Process ID and its information
+#[derive(Debug, Clone, Default)]
+pub struct ProcessInformation {
+    pub pid: usize,
+    pub cmdline: String,
+
+    inner_status: String,
+    inner_stat: String,
+
+    /// Processed `/proc/self/status` file
+    cached_status: Option<Rc<HashMap<String, String>>>,
+    /// Processed `/proc/self/stat` file
+    cached_stat: Option<Rc<Vec<String>>>,
```

Comment:
Why do you store both the original and the processed values?

Reply:
I was thinking of making a snapshot of all `/proc/<pid>` when the `pgrep` command is started, as there are times when slower processing may spawn new processes or some processes exit. Is this the right way to think about it?

Reply:
Yes, that makes sense.